### PR TITLE
configurator healthmonitor

### PR DESCRIPTION
**Pull request by pkharat**

### CODE REVIEW

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 1]
```diff
@@ -6,9 +6,15 @@
 from gbpservice.nfp.core import main
 import os
 from gbpservice.nfp.configurator.lib import utils
-
+from gbpservice.nfp.core import poll as nfp_poll
 LOG = logging.getLogger(__name__)
 
+SUCCESS = 'SUCCESS'
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Some of these constants are common to base driver and generic config. Why not add these to constants.py file in lib directory.

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 2]
```diff
@@ -131,10 +171,14 @@ def _get_driver(self, service_type):
         return self.drivers[service_type]()
 
     def _notification(self, data):
+        LOG.info(" Sending notification=%s " % (data))
         event = self.sc.new_event(
             id='NOTIFICATION_EVENT', key='NOTIFICATION_EVENT', data=data)
         self.sc.poll_event(event)
 
+    def handle_poll_event(self, ev):
+        pass
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Will handle_poll_event not called when a poll event is enqueued?

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 3]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
-            else:
-                data = {
-                        'context': context,
-                        'resource': resource,
-                        'request_info': request_info,
-                        'result': result
-                        }
-                notification_data['kwargs'].append(data)
+            import traceback
+            traceback.print_exc()
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Remove the traceback calls.

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 4]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
+            LOG.error("Failed to process ev.id=%s, ev=%s reason=%s" % (ev.id,
+                                                                       ev.data,
+                                                                       err))
+            result = FAILED
+
+        if ev.id == const.EVENT_CONFIGURE_HEALTHMONITOR:
+            if (kwargs.get('periodicity') == INITIAL and
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Need to cover the case where initial health check fails.

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 5]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
+                        self._notification(notification_data)
+                elif result == SUCCESS:
+                    """set fail_count to 0 if it had failed earlier even once
+                    """
+                    kwargs['fail_count'] = 0
+        elif ev.id == const.EVENT_CLEAR_HEALTHMONITOR:
+            """Stop current poll event. event.key is vmid which will stop
+               that particular service vm's health monitor
+            """
+            notification_data = self._prepare_notification_data(ev, result)
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Should we handle the case where the poll_event_done() is already invoked as a result of consecutive health check failures and again configurator gets clear_healthmonitor() API.  What will happen if poll_event_done() gets invoked for an unexisting event?

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 6]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
+                    'resource': resource,
+                    'request_info': request_info,
+                    'result': result
+                   }
+            notification_data['kwargs'].append(data)
+        return notification_data
+
+    def poll_event_cancel(self, ev):
+        LOG.error('Poll event cancelled for event %s ' % (ev.data))
+        result = FAILED
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Why do we need to send failed notification when poll event is canceled? The event can be canceled in positive cases also like clear_healthmonitor() API invocation right?

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 7]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
+        # processing function. To keep the context unchanged, delete the
+        # notification_data before invoking driver API.
+        notification_data = context.get('notification_data')
+        del context['notification_data']
+        resource = context.get('resource')
+        del context['resource']
+
+        msg = {'receiver': const.ORCHESTRATOR,
+               'resource': resource,
+               'method': ev.id,
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Event ID is in capitals. Method name should be in small.

File: gbpservice/nfp/configurator/agents/generic_config.py [hunk 8]
```diff
@@ -187,34 +235,97 @@ def handle_event(self, ev):
...
+            """
+            notification_data = self._prepare_notification_data(ev, result)
+            self._sc.poll_event_done(ev)
+            self._notification(notification_data)
+        else:
+            """For other events, irrespective of result send notification"""
+            notification_data = self._prepare_notification_data(ev, result)
             self._notification(notification_data)
 
+    def _prepare_notification_data(self, ev, result):
```
Comment (njagadish) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Add function headers for all new functions.

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 2 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: First decorator methods for poll event will be called and at the last this function gets called. Will remove this function

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 3 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Yeah, forgot to remove this.

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 4 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: In handle_event() method we are giving max_times arg for initial type of health monitor. After max_times attempts if the ev object is still being polled then poll_event_cancel() will be invoked by framework and this is also indication that initial health check failed. And here we are sending health check failed notification

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 5 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Framework should handle this, we should not keep track of ev objects right ?

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 7 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: okay

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 1 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: okay. Will add it.

Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Added generic_config agent needed constants in generic_config_constants.py and common constants in constants.py

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 7 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: When crossed checked this with the demo_26thfeb2016 branch. method name is same as ev.id. What should we do here ?

File: gbpservice/nfp/configurator/agents/generic_config.py (hunk 5 above)
Comment (pkharat) [inline gbpservice/nfp/configurator/agents/generic_config.py]: Verified this by trying to invoke poll_event_done for not existing poll event. Framework handles this gracefully

**Comment by pkharat** [comment]
Closing this pull request. New one is generated to merge into configurator-module branch

**Pull request merged**